code/modules/reagents/chemistry/machinery/chem_separator.dm

```diff
@@ -0,0 +1,287 @@
+		var/mutable_appearance/filling = mutable_appearance(icon_state, fill_name)
+		. += filling
+
+/// Checks whether the item can ignite the separator
+/obj/structure/chem_separator/proc/ignite_with(obj/item/object, mob/living/user)
+	var/ignition_message = object.ignition_effect(src, user)
+	if(!ignition_message)
+		return FALSE
+	user.visible_message(ignition_message)
+	fire_act(object.get_temperature())
+	return TRUE
+
+/obj/structure/chem_separator/attackby(obj/item/item, mob/user, params)
+	if(ignite_with(item, user))
+		return TRUE // no afterattack
+	if(is_reagent_container(item) && !(item.item_flags & ABSTRACT) && item.is_open_container())
+		var/obj/item/reagent_containers/new_beaker = item
+		if(!user.transferItemToLoc(new_beaker, src))
+			return FALSE
+		replace_beaker(user, new_beaker)
+		balloon_alert(user, "added beaker")
+		update_appearance(UPDATE_ICON)
+		return TRUE // no afterattack
+	return ..()
+
+/// Insert, replace or eject the container depending on the state and parameters
+/obj/structure/chem_separator/proc/replace_beaker(mob/living/user, obj/item/reagent_containers/new_beaker)
+	if(!user)
+		CRASH("[user] ([user?.type]) is not a living, but is trying to replace beaker")
+	if(beaker)
+		if(burning)
+			stop()
+		if(!issilicon(user) && in_range(src, user))
+			user.put_in_hands(beaker)
+			playsound(src, 'sound/items/handling/drinkglass_pickup.ogg', PICKUP_SOUND_VOLUME, ignore_walls = FALSE)
+		else
+			beaker.forceMove(drop_location())
+		beaker = null
+	if(new_beaker)
+		beaker = new_beaker
+		playsound(src, 'sound/items/handling/drinkglass_drop.ogg', DROP_SOUND_VOLUME, ignore_walls = FALSE)
+	update_appearance(UPDATE_ICON)
+	return TRUE
+
+/obj/structure/chem_separator/fire_act(exposed_temperature, exposed_volume)
+	if(!burning)
+		start()
+		return
+	return ..()
+
+/obj/structure/chem_separator/extinguish()
+	if(burning)
+		stop()
+	return ..()
+
+/// Ignite the burner to start the separation process
+/obj/structure/chem_separator/proc/start()
+	if(!beaker)
+		return
+	if(beaker.reagents.total_volume >= beaker.reagents.maximum_volume)
+		return
+	if(!reagents.total_volume)
+		return
+	var/list/reagents_sorted = reagents.reagent_list.Copy()
+	reagents_sorted = sort_list(reagents_sorted, GLOBAL_PROC_REF(cmp_reagents_asc))
+	separating_reagent = reagents_sorted[1]
+	burning = TRUE
+	update_appearance(UPDATE_ICON)
+	START_PROCESSING(SSobj, src)
+
+/// Extinguish the burner to stop the separation process
+/obj/structure/chem_separator/proc/stop()
+	separating_reagent = null
+	burning = FALSE
+	if(boiling)
+		boiling = FALSE
+		condenser.trans_to(reagents, condenser.total_volume)
+		soundloop.stop()
+	update_appearance(UPDATE_ICON)
+	STOP_PROCESSING(SSobj, src)
+
+/// Fill internal storage with reagents from the container
+/obj/structure/chem_separator/proc/load()
+	if(burning)
+		return
+	if(!beaker)
+		return
+	if(!beaker.reagents.total_volume)
```

Comment:
```suggestion
	if(!beaker?.reagents.total_volume)
```
This is not something you have to do, but it's a thing you can use to condense your checks